Add basic support for VS2017

siprbaum opened the pull request.

siprbaum commented:
This is best reviewed each commit individually. I also put a lot of information in the commit messages

siprbaum commented on src/GitTfs.VsCommon/TfsHelper.Vs2017Base.cs:
```diff
@@ -0,0 +1,201 @@
+                if (state.HasFlag(InstanceState.Local) && state.HasFlag(InstanceState.Registered))
+                {
+                    myVisualStudioInstallationPath = instance.GetInstallationPath();
+                    Trace.TraceInformation("Found matching Visual Studio version at {0}", myVisualStudioInstallationPath);
```
This currently is printed to often, e.g. for each `tfs-remote` in the .git/config file.
The underlying issue is that the GitTfs.Vs2017\TfsHelper.cs is created multiple times and therefore its base class (this file) is constructed multiple times.

Would it be possible to make this class a singleton? That whould mean to configure it on the DI container as such, but
I am little nervous about the consequences why this wasn't done before already.

Having it recreated for each git-tfs-remote would also mean we do this detection each and every time again ;-(

pmiossec replied:
Yes, I don't know the consequences either 😢 
Perhaps storing the path in a static field could be a not so bad solution...

siprbaum replied:
I had exactly what you proposed already here - or at least a common static state which should be shared over all instance,, but as this is a base class for theVS2017 TfsHelper and the (yet to come) VS2019 TfsHelper I figured it is not practical.
It would be a possibility to do it like here https://stackoverflow.com/a/32107465, e.g. make the state depend on the concrete type of the class.
 
But then I figured I should ask here first on your thoughts about why the TfsHelper instances are not made a singleton in the structuremap DI container.

I did make the modifcations to make it singleton locally  and used to import a few changesets, but then - because I was unsure e.g. about sharing between threads and other implications - didn't follow through with it.

pmiossec replied:
>It would be a possibility to do it like here https://stackoverflow.com/a/32107465, e.g. make the state depend on the concrete type of the class, but this also feels a little clumsy.

You could do that but instead of the class name, use the `majorVersion` as the key.

I would go with this idea because I also don't know the consequences to put `TfsHelper` instances as singleton...

siprbaum replied:
Yes, of course that would be better. 
I will try to implement this and add it to this PR, otherwise you get a lot of console outputs in case you interact with multiple TFS remotes. Currently, one line is written for each TFS branch we are dealing with, which is quite noisy!

spraints replied:
@siprbaum I don't recall any particular reason for instantiating the plugin multiple times. IIRC it's that way because it's the default in structure map. I took a quick look through the code and don't see any state that would be problematic to have as a singleton. I think a singleton would be better than storing things in a static var on an abstract class. :+1: on marking the tfs helper as a singleton.

siprbaum replied:
@spraints OK, I will look into it, making it a singleton if you think it is clearer. (I think this too, but currently it feels riskyier,
because this would also affect the working VS2015 support if there are unforseen sideeffects, which I tried to avoid till now)

siprbaum replied:
Added a commit to make the TfsHelper a singleton

siprbaum commented:
@pmiossec Did you have chance to look at the commits already? For me, the only open issue is the one about the quite noisy logging right now. This can either be solved via
1. Tone down this logging so it only appears in the trace file
2. Introduce a static class type specific member to collect the data which shall be shared over all TfsHelpers of the same type (e.g. the VS2017 TfsHelpers)
3. Configure the TfsHelper to be a singleton in the DI container so we don't create it multiple times

Right now I am leaning to point 2, as it would be quite a small change and probably only touches the last two commits
('Add minimal VS2017 support VS2017 using a hardcoded location ...' , 'Add support for VS2017')
So I would be interested if the prior commits are structured correctly, e.g. split in such a way they are understandable or if they need further refinements or if I should tackle this differently to enable easier review

pmiossec commented:
>Did you have chance to look at the commits already? 

Just a very light look to see if something evident stand out.
I have to spend more time on it (but time is something I lack at the moment 😢  )

Note: You should rebase to solve a conflict. I don't know if you want to do it now or later...

siprbaum commented:
Thx for the int, I rebased the PR to solve the conflict in the doc/release-notes/NEXT.md file

pmiossec reviewed (approved):
Working for me.
I saw nothing really disturbing and that cannot be fixed later.
Seems good.
Commit are very detailed 👍 

Note: I have pushed a small commit on top of your PR to fix a detail...

spraints commented on src/GitTfs.VsCommon/TfsHelper.Vs2017Base.cs:
```diff
@@ -61,7 +72,51 @@ protected override bool HasWorkItems(Changeset changeset)
+                }
+                else
+                {
+                    Trace.TraceWarning("Ingoring incomplete Visual Studio version at {0}", instance.GetInstallationPath());
```
Nit: typo on "Ingoring" -> "Ignoring".

siprbaum replied:
Typo fixed in the seprate commit 85cba621b35e60dacf605efc2ed511a8f5487751

spraints commented on src/GitTfs.VsCommon/TfsHelper.Vs2017Base.cs:
```diff
@@ -61,7 +72,51 @@ protected override bool HasWorkItems(Changeset changeset)
+                if (fetched <= 0)
+                    break;
+
+                var instance = (ISetupInstance2)instances[0];
```
Are there cases where this cast would fail?

siprbaum replied:
Not that I am aware of. But if it does, AFAIR we fail to instantiate the TfsHelper which would result in skipping the concrete plugin, e.g. we would think that VS2017 support is working.

spraints commented on src/GitTfs.VsCommon/TfsHelper.Vs2017Base.cs:
```diff
@@ -61,7 +72,51 @@ protected override bool HasWorkItems(Changeset changeset)
+                    continue;
+                }
+
+                if (myVisualStudioInstallationPath != null)
+                {
+                    Trace.TraceWarning("Already found a Visual Studio version. Ignoring version at {0}", instance.GetInstallationPath());
+                    continue;
+                }
```
Why is it still iterating at this point?

siprbaum replied:
I fought with myself here when implementing this, but then decidded to continue to iterate so we will at least show
**all** the  editions & versions of one VS major, e.g. all VS2017 instances on the machine.
Because at least I didn't find any garantuees if the order of the returned intances is always deterministic, e.g. could be that one git-tfs call will find VS2017 Enterprise whereas the next run will find the VS2017 Professional which is also installed, which is problematic in case of (yet to come) checkin policies, because there can be different policies installed in the different versions.

To sum it up, in case of multiple versions with the same major, I wanted to give the user a chance to see which one
is taken and which other versions of the same major are potentially there but are not used, as I explicitley stated in the commit messages that supporting multiple versions of the same major is out of scope right now.

spraints replied:
> To sum it up, in case of multiple versions with the same major, I wanted to give the user a chance to see which one is taken and which other versions of the same major are potentially there but are not used, as I explicitley stated in the commit messages that supporting multiple versions of the same major is out of scope right now.

This seems like a pretty rare edge case, where there are multiple versions and one of them won't work and git-tfs chooses the one that won't work. What do you think about simplifying this so that it stops looping after it finds a working installation? If it does end up being a problem, this can be changed later to let the user choose an install.

Alternatively, maybe reduce the log level to 'Info' so that it's not as noisy in benign cases.

siprbaum replied:
For checkin policies it is not a rare case, because at least here it happened to me that I simply had forgotten to install the checkin policies in both VS2017 versions and due to bad luck git-tfs has chosen to use the one which didn't have it installed. And in that case it was quite confusing why it didn't work.
But I agree with you that the rest of the functionality shouldn't care which VS2017 edition (or minor) is used. 

So the additional looping (and therefore the logging) only has an effect if more than one version of the **same** studio major is installed.  I think that having mutliple VS2017 instances installed is already a rare case , but I wanted to provide at least some breadcrumbs in case something goes wrong,  as it is otherwise not easy to figure out.

All others which have e.g. only one VS2019 and one VS2017 installed should not see any additional logging output, as that case is alread covered  with this comparision in the loop
```csharp
           if (version.Major != myMajorVersion)
           {
                    continue;
           }
```
At least that was my intention when writing the code.

spraints replied:
👍 I'll defer to you on this.

siprbaum replied:
Marked as resolved as I think it has value to show all versions, as explained above

siprbaum commented on src/GitTfs/Core/TfsInterop/TfsPlugin.cs:
```diff
@@ -23,7 +23,8 @@ public static TfsPlugin Find()
                        pluginLoader.Fail("Unable to load TFS version specified in GIT_TFS_CLIENT (" + explicitVersion + ")!");
             }
-            return pluginLoader.TryLoadVsPluginVersion("2015", true) ??
+            return pluginLoader.TryLoadVsPluginVersion("2017", true) ??
+                   pluginLoader.TryLoadVsPluginVersion("2017") ??
                    pluginLoader.TryLoadVsPluginVersion("2015") ??
```
Hm. I am not sure this logic works for VS2017 or VS2019, as the check for "is corresponding VS version installed" checks
for the existence of the registry entries
```
HKLM\SOFTWARE\Wow6432Node\Microsoft\DevDiv\vs\Servicing\<versionCode>
HKLM\SOFTWARE\Microsoft\DevDiv\vs\Servicing\<versionCode>
```
where `versionCode` would be 15.0 for VS2017.
But for e.g. VS2019 both keys would not be found with versionCode=16.0, because to the best of my knowledge, there is nothing there which could indicate that VS2019 is even installed.
For VS2017, I have the key, but I am not sure if this is by accident only. Could someone pls double check if either one of those keys exist for a VS2017 installation?


Currently, I am leaning to remove the optional isVisualStudioRequired parameter for `public TfsPlugin TryLoadVsPluginVersion(string version, bool isVisualStudioRequired = false)` altoegether, especially as it will not work for VS2019 anyhow.

spraints replied:
I think removing `isVisualStudioRequired` would make sense.

The intent of the current code is to try to automatically figure out which version of VS you're actually using on your computer, so that we are running in an env where you have checkin policies configured. (In the ancient past, TFVC assemblies weren't distributed with git-tfs, so git-tfs needed to probe around to see which version it was possible to load at all.)

All of this would probably make sense to replace with configuration. IOW, git-tfs could default to using the newest version unless you override it with `git config --local git-tfs.client-version VS2015` or `git config --global git-tfs.client-version VS2015`. I'd be :+1: on removing support for the `GIT_TFS_CLIENT` env var.

How does that sound?

siprbaum replied:
I already started working on removing `isVisualStudioRequired` yesterday, anticipating your answer 😄 

But I have currently no intention to implement the config based approach of selecting the version, at least not in this PR,
as it is orthogonal to the PRs goal and we already have the established and documented way to select a different VS version with the `GIT_TFS_CLIENT` environment variable. 
Don't get me wrong, I **like** what you are proposing, but I don't want to overextend the scope of this PR

spraints replied:
> But I have currently no intention to implement the config based approach of selecting the version, at least not in this PR,
> as it is orthogonal to the PRs goal and we already have the established and documented way to select a different VS version with the `GIT_TFS_CLIENT` environment variable.
> Don't get me wrong, I **like** what you are proposing, but I don't want to overextend the scope of this PR

:+1: :+1:

siprbaum commented:
I fixed the typos and added a commit which makes the TfsHelper a singleton.
So there is IMHO only the one unresolved comment from @spraints, which I am leaning to do differently as this will not work for VS2019 anyhow

spraints commented on src/GitTfs/Core/TfsInterop/TfsPlugin.cs:
```diff
@@ -161,6 +161,9 @@ public virtual void Initialize(IAssemblyScanner scan)
         {
+            // Mark the ITfsHelper as a singleton to ensure that we create it only once.
+            // Otherwise, it is created e.g. for every remote which is wasteful.
+            config.For<ITfsHelper>().Singleton();
```
Other singletons are marked with `[StructureMapSingleton]`. Can the plugins be marked that way instead?

siprbaum replied:
That is exactly what I tried first, but the current code only looks at that attribute of the assembly where the Program.cs is loated in, see
https://github.com/git-tfs/git-tfs/blob/1123ad84a8064540bf98faa01a8d1dfeae97d9cb/src/GitTfs/Program.cs#L178
It doesn't look at code implemented in other assemblies.

AFAIK the usage of the attribute is only for code where we actually also have one real implementation, e.g. like the [Globals](https://github.com/git-tfs/git-tfs/blob/master/src/GitTfs/Globals.cs#L13).
But for the `ITfsHelper`, we actually have multiple implementations, so I would prefere to keep it as is.

spraints replied:
Ah, right. 👍

spraints reviewed (approved):
👍 This is in good shape. I'd be happy to have this merge with or without the `isVisualStudioRequired` change.

siprbaum commented:
Scratch that last commit, It fails the unit tests, at least locally. Not sure why appveyor didn't fail

siprbaum commented:
I force pushed to fix the unit tests in the last commit.

siprbaum commented:
From my end, this would be good to go in
